sys/compat/linuxkpi/common/src/linux_compat.c

```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
 		uaddr_val += (uintptr_t)pts->bsd_ioctl_data;
 
 		/* update address location */
-		*uaddr = (void *)uaddr_val;
+		*uaddr = (void * __capability)(uintcap_t)uaddr_val;
 		return (1);
 	}
 	return (0);
 }
 
 int
-linux_copyin(const void *uaddr, void *kaddr, size_t len)
+linux_copyin(const void * __capability uaddr, void *kaddr, size_t len)
 {
-	if (linux_remap_address(__DECONST(void **, &uaddr), len)) {
+	if (linux_remap_address(__DECONST(void * __capability *, &uaddr), len)) {
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(kaddr, uaddr, len);
+		memcpy(kaddr, (void *)(__cheri_addr uintptr_t)uaddr, len);
```

Comment:
This would fail in the purecap kernel. Should probably use `(__cheri_fromcap void *)uaddr` that keeps `uaddr` from becoming untagged in pruecap mode.

Reply:
If it has not changed or has some special meaning with `uintptr_t`, `__cheri_addr` should not be ignored in purecap as I recall explicitly using it some places.

Reply:
But uintptr_t is uintcap_t in purecap, not vaddr_t.  __cheri_addr for a vaddr_t cast makes sense still, but I would expect a uintcap_t cast of a capability pointer to always be a no-op since you aren't losing bits unlike uintptr_t in hybrid (which is vaddr_t) where you are only getting 64 bits as a result of the cast.

Reply:
I just checked what the current compiler does and it crashes when casting a pointer to `(__cheri_addr __intcap_t)`, I do agree that it should mirror `__cheri_to/fromcap` and be a no-op in this case, otherwise it is confusing. I'm opening a clang issue about this.

Reply:
This now uses `__cheri_fromcap`